### ISSUE TITLE
Make remote tracebacks easier to read and distinguish

### ISSUE THREAD

**Issue by gjoseph92** [dask/distributed#4880]
When an exception occurs on the cluster, the traceback that shows up is long, and includes both a local component and a remote component. I suspect this is a stumbling block for beginners because:
1. When tracebacks are too long, people glaze over and don't read them
2. For those that do read them, separating the local client boilerplate from the remote part requires practice

Specifically, I always skim for something like `raise exception.with_traceback(traceback)` in `distributed/client.py`, ignore everything above it, and just look at the remote part of the traceback. Could we format our error messages differently so users don't have to learn this unintuitive skill?

Goals:
1. Users can easily tell whether an exception happened locally or on the cluster
2. Minimal internal distributed code is shown in tracebacks when we know the error wasn't an internal distributed error.

We already do a good job with 2 on the worker side thanks to [`get_traceback`](https://github.com/dask/distributed/blob/4f83686b9ffed4e8c41bbf002c1bab24b38b9dab/distributed/utils.py#L732-L744), which removes irrelevant frames. So this might be as simple as `raise ... from None` on the client when re-raising a remote exception. Plus somehow adding a prefix like:

```
------------------------------------------------------------------------------------------------
ValueError (remote)                               Traceback from cluster (most recent call last)

While calling Client.compute, this error occurred on worker 'worker-abcde'
while executing task ('map-blocks-12345', 0, 0):
```
so that remote exceptions are easily distinguishable from local ones.

**Comment by mrocklin** [comment]
There are some cases where I or others want to see the full traceback, but I agree that this probably shouldn't be the default.  I'd be comfortable cutting off the traceback at a certain point.  Ideally this would be controllable by a configuration value.

**Comment by gjoseph92** [comment]
I think the key is "when we _know_ the error wasn't an internal distributed error". Specifically, if the client is re-raising an exception sent from the cluster, I think we know that the preceding client frames aren't that relevant (besides maybe knowing the top-level client function, hence the `While calling Client.compute`).

> Ideally this would be controllable by a configuration value.

Also agreed.

**Comment by j-bennet** [comment]
@jrbourbeau @fjetter 

TL;DR:

* `raise... from None` won't solve the problem
* but manipulating traceback before raising the exception might
* there are several different ways to do it, we need to pick one.

I experimented a bit on a branch [here](https://github.com/dask/distributed/compare/main...j-bennet:j-bennet/4880-cleanup-remote-tracebacks?expand=1). Here are some findings.

1. `raise ... from None` is not helpful in this case, since in `distributed.Client`, we are not raising a new Exception, but re-raising the original exception, already with specific traceback (`raise exc.with_traceback(tb)`).
2. before re-raising the exception, we can remove more of the irrelevant traceback frames from traceback, however, this will only help partially. Some irrelevant frames will be pushed into the trace stack after `Client` raised. Here is an example unit test in my branch:

```
def throws_map_blocks(x, block_id=None):
    if block_id is not None and block_id[0] % 2 == 0:
        raise ValueError(f"Bad block id {block_id}")
    return x / 2


@pytest.mark.parametrize("clean_traceback,expected_levels", [(True, 9), (False, 16)])
def test_map_blocks_error(client, clean_traceback, expected_levels):
    da = pytest.importorskip("dask.array")
    arr = da.arange(30, chunks=3)
    op = arr.map_blocks(throws_map_blocks)
    try:
        with dask.config.set(
            {"distributed.exceptions.clean_traceback": clean_traceback}
        ):
            op.compute()
    except Exception:
        _, _, tb = sys.exc_info()
        traceback.print_tb(tb)
```

The traceback before cleanup has 16 stack frames:

<details>
<summary>Click me!</summary>
<pre>
  File "/Users/jbennet/src/distributed/distributed/tests/test_as_completed.py", line 196, in test_map_blocks_error
    op.compute()
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py", line 314, in compute
    (result,) = compute(self, traverse=False, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py", line 599, in compute
    results = schedule(dsk, keys, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 3229, in get
    raise exc_value.with_traceback(exc_traceback)
  File "/Users/jbennet/src/distributed/distributed/client.py", line 3225, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 2358, in gather
    return self.sync(
           ^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/utils.py", line 362, in sync
    return sync(
           ^^^^^
  File "/Users/jbennet/src/distributed/distributed/utils.py", line 430, in sync
    raise exc.with_traceback(tb)
  File "/Users/jbennet/src/distributed/distributed/utils.py", line 402, in f
    result = yield future
             ^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/tornado/gen.py", line 767, in run
    value = future.result()
            ^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 2221, in _gather
    raise exception.with_traceback(traceback)
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/optimization.py", line 990, in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
    ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/core.py", line 149, in get
    result = _execute_task(task, cache)
    ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
    ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/array/core.py", line 523, in _pass_extra_kwargs
    return func(*args[len(keys) :], **kwargs)
      ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/tests/test_as_completed.py", line 185, in throws_map_blocks
    raise ValueError(f"Bad block id {block_id}")
    ^^^^^^^^^^^^^^^^^]
</pre>
</details>

The traceback after cleanup has 6 stack frames:

<details>
<summary>Click me!</summary>
<pre>
  File "/Users/jbennet/src/distributed/distributed/tests/test_as_completed.py", line 199, in test_map_blocks_error
    op.compute()
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py", line 314, in compute
    (result,) = compute(self, traverse=False, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py", line 599, in compute
    results = schedule(dsk, keys, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 3228, in get
    raise exc_value.with_traceback(exc_traceback)
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/array/core.py", line 523, in _pass_extra_kwargs
    return func(*args[len(keys) :], **kwargs)
      ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/tests/test_as_completed.py", line 186, in throws_map_blocks
    raise ValueError(f"Bad block id {block_id}")
    ^^^^^^^^^^^^^^^^^
</pre>
</details>

This is helpful, right? Except, you may notice that I cheated. I included some `dask` code into my list of "internal code":

![image](https://user-images.githubusercontent.com/637013/234131694-9b7683c2-99d1-4a2e-8e7f-f8312e82e8b8.png)

This brings me to my next finding.

3. There are a number of Python libraries that have to fudge their traceback. For example, [ipython](https://github.com/ipython/ipython/blob/47d043e1dd051294d85f41b18f8e3c1f8434fd5c/IPython/core/debugger.py#L22), [jinja2](https://github.com/pallets/jinja/blob/090a5a9e22a7888da93ead324b33b255d106a0c2/src/jinja2/debug.py#L51), [IDLE](https://github.com/python/cpython/blob/01be52e42eac468b6511b56ee60cd1b99baf3848/Lib/idlelib/run.py#L264).

They are using several different methods to mark code internal.

* Keep a list of internal code paths (optionally including function names), like I did here.
* Decorate internal function with something that will push it into a singleton list.
* Set a magic local value inside this frame, `ipython` sets `__tracebackhide__`.

2nd and 3rd are more explicit and feel more like "the right way" to do this. With the first, I can remove traceback frames that don't belong to `distributed` code, which is kind of nice, on one hand, but is probably too invasive.

**Comment by j-bennet** [comment]
In my branch [here](https://github.com/dask/distributed/compare/main...j-bennet:j-bennet/4880-cleanup-remote-tracebacks?expand=1), the same traceback would be quite a bit shorter.

```python
import dask
import dask.array as da
from distributed import Client

def throws_map_blocks(x, block_id=None):
    if block_id is not None and block_id[0] % 2 == 0:
        raise ValueError(f"Bad block id {block_id}")
    return x / 2

dask.config.set({'distributed.exceptions.clean_traceback': True})
client = Client(silence_logs="ERROR")
arr = da.arange(30, chunks=3)
op = arr.map_blocks(throws_map_blocks)

op.compute()
```

The output is:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 15
     12 arr = da.arange(30, chunks=3)
     13 op = arr.map_blocks(throws_map_blocks)
---> 15 op.compute()

File ~/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py:314, in DaskMethodsMixin.compute(self, **kwargs)
    290 def compute(self, **kwargs):
    291     """Compute this dask collection
    292
    293     This turns a lazy Dask collection into its in-memory equivalent.
   (...)
    312     dask.compute
    313     """
--> 314     (result,) = compute(self, traverse=False, **kwargs)
    315     return result

File ~/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py:599, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    596     keys.append(x.__dask_keys__())
    597     postcomputes.append(x.__dask_postcompute__())
--> 599 results = schedule(dsk, keys, **kwargs)
    600 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File ~/src/distributed/distributed/client.py:3237, in Client.get(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   3235     exc_type, exc_value, exc_traceback = sys.exc_info()
   3236     exc_traceback = drop_internal_traceback(exc_traceback)
-> 3237     raise exc_value.with_traceback(exc_traceback)
   3238 finally:
   3239     for f in futures.values():

Cell In[1], line 7, in throws_map_blocks()
      5 def throws_map_blocks(x, block_id=None):
      6     if block_id is not None and block_id[0] % 2 == 0:
----> 7         raise ValueError(f"Bad block id {block_id}")
      8     return x / 2

ValueError: Bad block id (2,)
```

So it's better, but it doesn't quite do what you want. This is because we can only do so much in `distributed`, before the exception bubbles up to `dask/base`. Dask frames will be added to the stack after the stack was filtered in `distributed.client`.
